Document Personal Access Token feature of API ML

taban03 opened the pull request:
Signed-off-by: at670475 <andrea.tabone@broadcom.com>

### Your checklist for this pull request
🚨Please review the [guidelines for contributing](../docs/contribute/contributing.md) to this repository.

- [ ] If the changes in this PR is part of the next future release, make this pull request against the **docs-staging** branch which will be published at the next release boundary. If the changes in this PR are part of the current release, use the default base branch, **master**. For more information about branches, see https://github.com/zowe/docs-site/tree/master#understanding-the-doc-branches. 
      
- [ ] If this PR relates to GitHub issues in `docs-site` or other repositories, please list in Description, prefixed with **close**, **fix** or **resolve** keywords.

### Description (including links to related git issues)
Please describe your pull request.

:heart:Thank you!

balhar-jakub commented on docs/extend/extend-apiml/api-mediation-personal-access-token.md:
```diff
@@ -0,0 +1,151 @@
+# Personal Access Token
```
I ould move it to use section instead of Extend.

balhar-jakub commented on docs/extend/extend-apiml/api-mediation-personal-access-token.md:
```diff
@@ -0,0 +1,151 @@
+# Personal Access Token
+
+The API ML allows generating, validating and invalidating a **Personal Access Token (PAT)** that can be used, for instance, to leverage modern tools such as Version Control Systems (VCS) during
```
That could help access the tools such as VCS without using the credentials of the specific person and storing the credentials in the automation.

balhar-jakub replied:
I would also mention here the possibility to limit access to specific services.

balhar-jakub commented on docs/extend/extend-apiml/api-mediation-personal-access-token.md:
```diff
@@ -0,0 +1,151 @@
+
+To do that, the API ML offers a set of REST APIs to:
+
+1. [Generate the token](#generate-of-the-token)
```
I would structure the APIs to ones used by User and ones used by the administrator.

balhar-jakub replied:
Like directly in this list.

balhar-jakub commented on docs/extend/extend-apiml/api-mediation-personal-access-token.md:
```diff
@@ -0,0 +1,151 @@
+* Invalidate all tokens
+* Evict invalidated tokens and rules which are not relevant anymore
+
+## Generate of the token
```
Generate the token or Generation of the token

balhar-jakub commented on docs/extend/extend-apiml/api-mediation-personal-access-token.md:
````diff
@@ -0,0 +1,151 @@
+}
+```
+
+The **validity** refers to the expiration time of the tokens. The maximum threshold is 90 days.
````
The maximum threshold is configurable, isn't it?

taban03 replied:
No, it is not configurable for the time being.

balhar-jakub commented on docs/extend/extend-apiml/api-mediation-personal-access-token.md:
```diff
@@ -0,0 +1,151 @@
+The **validity** refers to the expiration time of the tokens. The maximum threshold is 90 days.
+The **scopes** allows to limit the access on a service level, so it will introduce higher level of security in some aspects. Users will be forced to provide a scope and if no service is specified, it will work for none.
+
+When creation is successful, the response to the request is a body containing the PAT. When creation fails, the user receives a 401 status code.
```
It would be good to show an example of the successful and unsuccessful response in this case.

balhar-jakub commented on docs/extend/extend-apiml/api-mediation-personal-access-token.md:
```diff
@@ -0,0 +1,151 @@
+
+## Validate the token
+
+The user can validate the Personal Access Token by calling the following REST API endpoint via Gateway:
```
Maybe explain this is usually not needed as the API ML does the validation when it receives the token in the request.

balhar-jakub commented on docs/extend/extend-apiml/api-mediation-personal-access-token.md:
````diff
@@ -0,0 +1,151 @@
+}
+```
+
+When validation is successful, the response to the request is a an empty body. When validation fails, the user receives a 401 status code.
````
I guess the response will also have 200 status code?

balhar-jakub commented on docs/extend/extend-apiml/api-mediation-personal-access-token.md:
````diff
@@ -0,0 +1,151 @@
+}
+```
+
+When the `/auth/access-token/revoke` endpoint is called, the provided PAT's hash will be stored in the Caching Service will be storing only hashes of already invalidated tokens. Access to these entries is protected by API ML’s client certificate.
````
will be stored. The Caching Service will be storing

balhar-jakub commented on docs/extend/extend-apiml/api-mediation-personal-access-token.md:
```diff
@@ -0,0 +1,151 @@
+
+When the `/auth/access-token/revoke` endpoint is called, the provided PAT's hash will be stored in the Caching Service will be storing only hashes of already invalidated tokens. Access to these entries is protected by API ML’s client certificate.
+
+When invalidation is successful, the response to the request is an empty body.
```
with 200 status code. 

Also I would mention what happens when the response isn't successful. Is there any specific behavior?

balhar-jakub commented on docs/extend/extend-apiml/api-mediation-personal-access-token.md:
```diff
@@ -0,0 +1,151 @@
+
+## Invalidate all the tokens for a user
+
+In case of possibility of breach, the Security Administrator can invalidate all the tokens based on some criteria.
```
Maybe more like:

When there is a suspicion on breach

balhar-jakub commented on docs/extend/extend-apiml/api-mediation-personal-access-token.md:
```diff
@@ -0,0 +1,151 @@
+Such criteria define the level of access control and can restrict the access in advance. This can be done by using
+**rules**, and the access restriction can be applied by either user ID or service scopes.
+
+The administrator can invalidate all the tokens bound to a specific user by calling the following REST API endpoint via Gateway:
```
What does administrator mean? Does this account need to have specific SAF resources? Or how is it decided

taban03 replied:
I fixed it and specified that he should have saf resources access in order to call these endpoints

balhar-jakub commented on docs/extend/extend-apiml/api-mediation-personal-access-token.md:
````diff
@@ -0,0 +1,151 @@
+}
+```
+
+The `userId` is the user the revocation should be applied to, while the timestamp represents the current date of revocation, in milliseconds. 
````
What is the current date in miliseconds relevant for?

taban03 replied:
The timestamp represents the date of revocation (the default value is the current time), in milliseconds, and it is
used to state that the tokens created before the date specified in the timestamp are invalidated. Therefore, any other tokens created after that date will not be affected by the service rule. I will fix it.

balhar-jakub commented on docs/extend/extend-apiml/api-mediation-personal-access-token.md:
```diff
@@ -0,0 +1,151 @@
+
+## Invalidate all the tokens for a service
+
+As we mentioned in the above section, it is possible to revoke tokens also by using rules based on service scopes.
```
I would start with: The administrator can invalidate all the tokens bound to

balhar-jakub commented on docs/extend/extend-apiml/api-mediation-personal-access-token.md:
```diff
@@ -0,0 +1,151 @@
+
+## Retrieve all the tokens and rules
+
+The user can retrieve all the invalidated tokens and rules which were stored in the cache by calling the following REST API endpoint via Gateway:
```
Who has access to this endpoint?

taban03 replied:
It's not valid anymore, as the user should not retrieve the tokens and rules. It should not be documented.

balhar-jakub commented on docs/extend/extend-apiml/api-mediation-personal-access-token.md:
```diff
@@ -0,0 +1,151 @@
+
+The `serviceId` represents the service the revocation should be applied to (e.g. APPL IDs), while the timestamp represents the current date of revocation, in milliseconds.
+
+When invalidation is successful, the response to the request is an empty body.
```
Again I guess with 200 as status code?

balhar-jakub commented on docs/extend/extend-apiml/api-mediation-personal-access-token.md:
```diff
@@ -0,0 +1,151 @@
+
+The `/auth/access-token/evict` endpoint evicts the all the invalidated tokens which were expired and all the rules related to the expired tokens.
+
+The main purpose of the eviction API is to make sure that the size of the cache does not grow unbounded.
```
And this isn't due to the actual size on disk but for the verification that the tokens are valid all the rules needs to be consulted.

balhar-jakub reviewed (commented):
There are some comments. 

I am also missing information on configuration of the maximum validity time of the token. Was that removed from the scope?

taban03 commented:
> There are some comments.
> 
> I am also missing information on configuration of the maximum validity time of the token. Was that removed from the scope?

I've addressed all the comments, thanks for reviewing. Regarding the configuration for the validity time, it is not configurable right now.
